src/olink/core/protocol.h

```diff
@@ -23,62 +23,188 @@
+ * Contains handlers for all messages in protocol.
+ * Not all handlers have to have full implementation for the protocol listnener object,
+ * some messages should be handled by server side, some by client side.
  */
 class OLINK_EXPORT IProtocolListener
 {
 public:
-    virtual ~IProtocolListener();
+    virtual ~IProtocolListener() = default;
 
-    virtual void handleLink(std::string name) = 0;
-    virtual void handleUnlink(std::string name) = 0;
-    virtual void handleInit(std::string name, nlohmann::json props) = 0;
-    virtual void handleSetProperty(std::string name, nlohmann::json value) = 0;
-    virtual void handlePropertyChange(std::string name, nlohmann::json value) = 0;
-    virtual void handleInvoke(int requestId, std::string name, nlohmann::json args) = 0;
-    virtual void handleInvokeReply(int requestId, std::string name, nlohmann::json value) = 0;
-    virtual void handleSignal(std::string name, nlohmann::json args) = 0;
-    virtual void handleError(int msgType, int requestId, std::string error) = 0;
+    /**
+     * Server side handler, handles link message.
+     * @param objectId Id of an object for which link message was received.
+     */
+    virtual void handleLink(const std::string&objectId) = 0;
+    /**
+     * Server side handler, handles unlink message.
+     * @param objectId Id of an object for which unlink message was received.
+     */
+    virtual void handleUnlink(const std::string&objectId) = 0;
+    /**
+     * Client side handler, handles remote init message.
+     * @param objectId Id of an object for which Init message was received
+     * @param props Current values of propeties for object service.
+     */
+    virtual void handleInit(const std::string&objectId , const nlohmann::json& props) = 0;
+    /**
+     * Server side handler, handles setProperty message.
+     * @param propertyId Unambigiously describes property in object for which setProperty message was received.
+     * @param value A value to which client request a property to be set.
+     */
+    virtual void handleSetProperty(const std::string&propertyId , const nlohmann::json& value) = 0;
+    /**
+     * Client side handler, handles propertyChange message.
+     * @param propertyId Unambigiously describes property in object for which propertyChange message was received.
+     * @param value A current value of property on server side
+     */
+    virtual void handlePropertyChange(const std::string&propertyId , const nlohmann::json& value) = 0;
+    /**
+     * Server side handler, handles Invoke message.
+     * @param requestId Id of a invoke request, which should be sent back with the response.
+     * @param methodId Unambigiously describes method in object for which invoke message was received.
+     * @param args Arguments with which method should be invoked.
+     */
+    virtual void handleInvoke(int requestId, const std::string&methodId , const nlohmann::json& args) = 0;
```

Comment:
Is this some kind of modern c++? I think */& belongs to the type.

So `const std::string& methodId , const nlohmann::json& args`

Reply:
string*, string, string&
are 3different types, so you are correct, that the* or & belongs to type
if you type string - it means you will allocate new memory and make a copy of sth was pass there, so what you got will not change and is always there (in that scope_
the * and & mean that you will point to some memory that someone else manages - it may change or even be deleted at some point, the difference is that in case of * you're saying -"I allow that there is nothing in there" ( but then you should always check against null before usage) and in other case you're saying I believe that there is always something